### PR TITLE
Storage market protocols versions 1.1.0

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
deal status request signature check

Signed-off-by: Alexey-N-Chernyshov <alexey.n.chernyshov@gmail.com>

### Description of the Change
Deal protocols v1.1.1.
Request deal status check.

### CODE REVIEW

**Comment by turuslan** [comment]
Need to add `Named` encodable types for newer protocol versions.

lotus old: https://github.com/filecoin-project/go-fil-markets/blob/v1.13.1/storagemarket/migrations/migrations_cbor_gen.go#L1634-L1648
lotus new: https://github.com/filecoin-project/go-fil-markets/blob/v1.13.1/storagemarket/network/types_cbor_gen.go#L24-L51

c++ `Named` example: https://github.com/filecoin-project/cpp-filecoin/blob/master/core/markets/retrieval/protocols/retrieval_protocol.hpp#L30-L49

**Review by wer1st — APPROVED** [review]

File: core/markets/storage/mk_protocol.hpp [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+    /** Waiting for deal to appear on chain */
+    STORAGE_DEAL_PUBLISHING,
+
+    /** deal failed with an unexpected error */
+    STORAGE_DEAL_ERROR,
+  };
+
+  /** Base MinerDeal */
+  struct MinerDeal {
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:140]: there are more fields

https://github.com/filecoin-project/go-fil-markets/blob/v1.13.1/storagemarket/migrations/migrations.go#L49-L69

https://github.com/filecoin-project/go-fil-markets/blob/v1.13.1/storagemarket/types.go#L91-L115

File: core/markets/storage/mk_protocol.hpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    DataRefV1_0_1 ref;
+    s >> ref;
+    v.ref = ref;
+    s >> v.deal_id;
+    return s;
+  }
+
+  struct MinerDealV1_1_0 : public MinerDeal {};
+
+  inline CBOR2_ENCODE(MinerDealV1_1_0) {
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:185]: named

https://github.com/filecoin-project/go-fil-markets/blob/v1.13.1/storagemarket/types_cbor_gen.go#L719

File: core/markets/storage/mk_protocol.hpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    s >> v.message;
+    DataRefV1_1_0 ref;
+    s >> ref;
+    v.ref = ref;
+    s >> v.deal_id;
+    return s;
+  }
+
+  /** Base class */
+  struct ClientDeal {
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:210]: like MinerDeal, there are more fields in lotus

File: core/markets/storage/mk_protocol.hpp [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    DealId deal_id;
+  };
+
+  struct MinerDealV1_0_1 : public MinerDeal {};
+
+  inline CBOR2_ENCODE(MinerDealV1_0_1) {
+    s << v.client_deal_proposal << v.proposal_cid << v.add_funds_cid
+      << v.publish_cid << v.client << v.state << v.piece_path << v.metadata_path
+      << v.is_fast_retrieval << v.message << v.message << DataRefV1_0_1{v.ref}
+      << v.deal_id;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:161]: tuple and cast
```suggestion
    s << (s.list() << v.client_deal_proposal << v.proposal_cid << v.add_funds_cid
      << v.publish_cid << v.client << v.state << v.piece_path << v.metadata_path
      << v.is_fast_retrieval << v.message << v.message << static_cast<const DataRefV1_0_1 &>(v.ref)
      << v.deal_id);
```

File: core/markets/storage/mk_protocol.hpp [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    s >> v.client;
+    s >> v.state;
+    s >> v.piece_path;
+    s >> v.metadata_path;
+    s >> v.is_fast_retrieval;
+    s >> v.message;
+    DataRefV1_0_1 ref;
+    s >> ref;
+    v.ref = ref;
+    s >> v.deal_id;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:179]: tuple and cast
```suggestion
    auto l{s.list()};
    l >> v.client_deal_proposal
      >> v.proposal_cid
      >> v.add_funds_cid
      >> v.publish_cid
      >> v.client
      >> v.state
      >> v.piece_path
      >> v.metadata_path
      >> v.is_fast_retrieval
      >> v.message
      >> static_cast<DataRefV1_0_1 &>(v.ref)
      >> v.deal_id;
```

File: core/markets/storage/mk_protocol.hpp [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    ClientDealProposal deal_proposal;
+    DataRef piece;
+    bool is_fast_retrieval = false;
+  };
+
+  struct ProposalV1_0_1 : public Proposal {};
+
+  inline CBOR2_ENCODE(ProposalV1_0_1) {
+    return s << v.deal_proposal << DataRefV1_0_1{v.piece}
+             << v.is_fast_retrieval;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:265]: tuple and cast
```suggestion
    return s << (s.list() << v.deal_proposal << static_cast<const DataRefV1_0_1 &>(v.piece)
             << v.is_fast_retrieval);
```

File: core/markets/storage/mk_protocol.hpp [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+  }
+  inline CBOR2_DECODE(ProposalV1_0_1) {
+    s >> v.deal_proposal;
+    DataRefV1_0_1 piece;
+    v.piece = piece;
+    s >> piece;
+    s >> v.is_fast_retrieval;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:272]: tuple and cast
```suggestion
    auto l{s.list()};
    l >> v.deal_proposal >> static_cast<DataRefV1_0_1 &>(v.piece) >> v.is_fast_retrieval;
```

File: core/markets/storage/mk_protocol.hpp [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    return s;
+  }
+
+  struct ProposalV1_1_0 : public Proposal {};
+
+  inline CBOR2_ENCODE(ProposalV1_1_0) {
+    auto m{CborEncodeStream::map()};
+    m["DealProposal"] << v.deal_proposal;
+    m["Piece"] << DataRefV1_1_0{v.piece};
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:281]:
```suggestion
    m["Piece"] << static_cast<const DataRefV1_1_0 &>(v.piece);
```

File: core/markets/storage/mk_protocol.hpp [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    m["FastRetrieval"] << v.is_fast_retrieval;
+    return s << m;
+  }
+
+  inline CBOR2_DECODE(ProposalV1_1_0) {
+    auto m{s.map()};
+    CborDecodeStream::named(m, "DealProposal") >> v.deal_proposal;
+    DataRefV1_1_0 piece;
+    CborDecodeStream::named(m, "Piece") >> piece;
+    v.piece = piece;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:291]:
```suggestion
    CborDecodeStream::named(m, "Piece") >> static_cast<DataRefV1_1_0 &>(v.piece);
```

File: core/markets/storage/mk_protocol.hpp [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+  };
+
+  struct SignedResponseV1_0_1 : public SignedResponse {
+    SignedResponseV1_0_1() = default;
+    explicit SignedResponseV1_0_1(Response response) {
+      this->response = std::move(response);
+    }
+
+    outcome::result<Bytes> getDigest() const override {
+      return codec::cbor::encode(ResponseV1_0_1{this->response});
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:354]:
```suggestion
      return codec::cbor::encode(static_cast<const ResponseV1_0_1 &>(response));
```

File: core/markets/storage/mk_protocol.hpp [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    };
+  };
+
+  inline CBOR2_ENCODE(SignedResponseV1_0_1) {
+    return s << ResponseV1_0_1{v.response} << v.signature;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:359]: tuple and cast
```suggestion
    return s << (s.list() << static_cast<const ResponseV1_0_1 &>(v.response) << v.signature);
```

File: core/markets/storage/mk_protocol.hpp [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+  }
+  inline CBOR2_DECODE(SignedResponseV1_0_1) {
+    ResponseV1_0_1 response;
+    s >> response;
+    v.response = response;
+    s >> v.signature;
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:365]: tuple and cast

File: core/markets/storage/mk_protocol.hpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/mk_protocol.hpp:140]: MinerDeal and ClientDeal are internal provider states of the deals and not a part of the protocol. I'm going to move it away from protocol headers.

File: core/markets/storage/ask_protocol.hpp [hunk 13]
```diff
@@ -47,12 +58,105 @@ namespace fc::markets::storage {
...
+    m["Ask"] << StorageAskV1_1_0{v.ask};
+    m["Signature"] << v.signature;
+    return s << m;
+  }
+
+  inline CBOR2_DECODE(SignedStorageAskV1_1_0) {
+    auto m{s.map()};
+    StorageAskV1_1_0 ask;
+    CborDecodeStream::named(m, "Ask") >> ask;
+    v.ask = ask;
```
Comment (turuslan) [inline core/markets/storage/ask_protocol.hpp:156]:
```suggestion
    v.ask = CborDecodeStream::named(m, "Ask").get<StorageAskV1_1_0>();
```

File: core/markets/storage/mk_protocol.hpp [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    // given
+    UnpaddedPieceSize piece_size;
+    // Optional: used as the denominator when calculating transfer
+    uint64_t raw_block_size{};
+  };
+
+  /** For protocol v1.0.1. */
+  struct DataRefV1_0_1 : public DataRef {};
+
+  CBOR_TUPLE(DataRef, transfer_type, root, piece_cid, piece_size)
```
Comment (turuslan) [inline core/markets/storage/mk_protocol.hpp:61]:
```suggestion
  CBOR_TUPLE(DataRefV1_0_1, transfer_type, root, piece_cid, piece_size)
```

**Review by turuslan — APPROVED** [review]

**Pull request merged**